### PR TITLE
[quant][graphmode] InsertObservers for shared class types

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30553 [quant][graphmode] `insert_quant_dequant` pass support shared class types
* #30552 [quant][graphmode][refactor] Add registerQParams function
* #30551 [quant][graphmode] Insert GetAttr for quantization parameters instead of Constant
* #30550 [quant][graphmode] Make `InsertQuantDeQuantHelper` global
* #30549 [quant][graphmode] Rename QuantizeHelper to InsertQuantDeQuantHelper
* **#30548 [quant][graphmode] InsertObservers for shared class types**

Summary:
ClassTypes can be shared among different module instances, but previously we assumed
they would be unique, this PR enables the insert_observers pass to work with shared class types
Test Plan:
python test/test_jit.py
python test/test_quantization.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18802465](https://our.internmc.facebook.com/intern/diff/D18802465)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -1151,6 +1152,28 @@ def forward(self, x):
...
+        qconfig_dict = {'': script_qconfig(default_qconfig)}
+        torch._C._jit_pass_insert_observers(m._c, "forward", qconfig_dict, True)
+        # conv1 and conv2 shares the same type, we need to
+        # make sure we didn't quantize the type twice
+        assert len([x for x, _ in m.conv1._modules._c.items()
+                    if x.startswith('_observer_')]) == 3, \
+            'Expected to have 3 observer submodules'
+        assert len([x for x, _ in m.conv2._modules._c.items()
+                    if x.startswith('_observer_')]) == 3, \
+            'Expected to have 3 observer submodules'
```
Comment (ZolotukhinM) [inline test/test_jit.py]: Could we somehow check that the conv modules have the same type after quantization? Also, just to verify: the test does fail without this change, right?

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -421,6 +424,17 @@ void InsertObserversHelper::insertObservers(
...
+
+  if (graph_observer_map_.count(graph.get())) {
+    // instance clone of observer module and setAttr
+    for (const auto& observer_attrs : graph_observer_map_.at(graph.get())) {
+      const auto& name = std::get<0>(observer_attrs);
+      const auto& observer = std::get<1>(observer_attrs);
+      module._ivalue()->setAttr(name, observer.clone_instance()._ivalue());
+    }
+    return;
+  }
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This condition should probably be moved to the beginning of the function. There is no need to repeat this work for every invoked method.

**Review by ZolotukhinM — APPROVED** [review]
Looks good!

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -178,6 +178,8 @@ class InsertObserversHelper {
   int uid_ = 0;
   // Set of observer forward call nodes
   std::unordered_set<Node*> observer_nodes_;
+  // Map from type of the module to the vector of observer modules we added to the class
+  std::unordered_map<Graph*, std::vector<std::tuple<std::string, script::Module>>> graph_observer_map_;
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: oh it should be graph, I forgot to change the comment

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: what do you mean? 
We didn't repeat this part for every invoked method, this is outside of the loop of calling insertObservers for invoked methods.
this is intentionally put after we call insertObservers for all invoked methods.

File: test/test_jit.py (hunk 1 above)
Comment (jerryzh168) [reply test/test_jit.py]: yeah this fails before the change. I didn't find a way check the types are the same though

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: when I dump the module I can find:
   conv1 = <__torch__.torch.nn.modules.conv.Conv2d object at 0x560b745835a0>
    conv2 = <__torch__.torch.nn.modules.conv.Conv2d object at 0x560b745c6490>

**Pull request closed without merging**